src/interpreter/mod.rs

```diff
@@ -0,0 +1,492 @@
+                if cond.get_bool().map_err(error_mapper(dsym.clone()))? {
+                    if let Statement::Block(_, stmts) = *block.clone() {
+                        self.exec_step_block(&stmts)
+                    } else {
+                        unreachable!("semantic fail");
+                    }
+                } else {
+                    Ok(None)
+                }
+            }
+            IfThenElse(_, cond, block_true, block_false) => {
+                let cond = eval_expr(&self.cur_frame, cond.as_ref())?;
+                if cond.get_bool().unwrap() {
+                    if let Statement::Block(_, stmts) = *block_true.clone() {
+                        self.exec_step_block(&stmts)
+                    } else {
+                        unreachable!("semantic fail");
+                    }
+                } else if let Statement::Block(_, stmts) = *block_false.clone() {
+                    self.exec_step_block(&stmts)
+                } else {
+                    unreachable!("semantic fail");
+                }
+            }
+
+            Transition(_, id, block) => {
+                if let Statement::Block(_, stmts) = *block.clone() {
+                    self.exec_step_block(&stmts)?;
+                } else {
+                    unreachable!("semantic fail");
+                }
+
+                Ok(Some(id.name()))
+            }
+
+            SignalDecl(_, _) => Ok(None),
+            WGVarDecl(_, _) => Ok(None),
+            Block(_, stmts) => self.exec_step_block(stmts),
+            Assert(_, _) => Ok(None),
+            StateDecl(_, _, _) => Ok(None),
+        }
+    }
+
+    fn exec_assign(
+        &mut self,
+        id: &Identifier,
+        expr: &Expression<BigInt, Identifier>,
+    ) -> Result<(), Message> {
+        let value = eval_expr(&self.cur_frame, expr)?;
+        // TODO: add context to error message
+
+        self.cur_frame.set_value(id, &value);
+
+        Ok(())
+    }
+
+    fn find_state_decl(
+        &mut self,
+        machine_dsym: &DebugSymRef,
+        machine_block: &[Statement<BigInt, Identifier>],
+    ) -> Option<Statement<BigInt, Identifier>> {
+        for stmt in machine_block {
+            if let Statement::StateDecl(_, state_id, _) = stmt {
+                if *state_id.name() == self.cur_frame.get_state() {
+                    return Some(stmt.clone());
+                }
+            }
+        }
+
+        // final state can be omited
+        if self.cur_frame.get_state() == "final" {
+            Some(Statement::StateDecl(
+                machine_dsym.clone(),
+                self.cur_frame.get_state().into(),
+                Box::new(Statement::Block(machine_dsym.clone(), vec![])),
+            ))
+        } else {
+            None
+        }
+    }
+}
+
+pub fn run<F: Field + Hash>(
+    ast: &[TLDecl<BigInt, Identifier>],
+    symbols: &SymTable,
+    mapping: &SymbolSignalMapping,
+    input: HashMap<String, F>,
```

Comment:
Same here, is input a mapping of signals to values?

Reply:
I don't understand this question, it is a mapping of identifiers to values. Do you mean that it should be commented?

Reply:
Yes